src/libp2p_crypto.erl

```diff
@@ -370,10 +500,218 @@ base58check_decode(B58) ->
+    Keys = lists:map(fun pubkey_to_bin/1, Keys0),
+    ISigsOrd = lists:sublist([{I, S} || {I, {_, S}} <- IKeySigs], M),
+    ISigsShuf = list_shuffle(ISigsOrd),
+    {ok, KeysDigest} = multihash:digest(iolist_to_binary(Keys), HashType),
+    {ok, MultiPubKey} = make_multisig_pubkey(M, N, Keys0, HashType),
+    {ok, MultiSigOrd} = make_multisig_signature(Msg, MultiPubKey, Keys0, ISigsOrd),
+    {ok, MultiSigShuf} = make_multisig_signature(Msg, MultiPubKey, Keys0, ISigsShuf),
+
+    MsgBad = <<Msg/binary, "Please help me I'm really in need of HNT at the moment">>,
+    MultiSigOrdBad = <<MultiSigOrd/binary, "looks legit">>,
+    MultiPassBad = {multisig, M, N, <<KeysDigest/binary, "hi">>},
+
+    %% Reorder keys
+    KeysReord =
+        (fun() ->
+            case Keys of
+                [K1, K2 | Ks] -> [K2, K1 | Ks];
+                Keys -> Keys
+            end
+        end)(),
+    MultiSigKeysReord =
+        iolist_to_binary([KeysReord, ISigsToBin(ISigsOrd)]),
+
+    %% Break index on a random isig, by pushing it out of range:
+    MultiSigOrdWithOutOfRangeIndex =
+        (fun () ->
+            R = rand:uniform(M) - 1, % Correct for 0-index
+            Replace =
+                fun ({I, S}) when I =:= R -> {N + 1, S}; (IS) -> IS end,
+            ISigsOutOfRange = lists:map(Replace, ISigsOrd),
+            iolist_to_binary([Keys, ISigsToBin(ISigsOutOfRange)])
+        end)(),
+
+    %% Reuse same signature M times
+    MaybeClonedSigTest =
+        (fun() ->
+            case M > 1 of
+                false ->
+                    [];
+                true ->
+                    [IS | _] = ISigsOrd,
+                    Clones = lists:duplicate(M, IS),
+                    MultiSigOrdFromClones =
+                        iolist_to_binary([Keys, ISigsToBin(Clones)]),
+                    [?_assertNot(verify(Msg, MultiSigOrdFromClones, MultiPubKey))]
+            end
+        end)(),
+
+    [
+        %% === Good:
+        ?_assert(verify(Msg, MultiSigOrd, MultiPubKey)),
+        ?_assert(verify(Msg, MultiSigShuf, MultiPubKey)),
+
+        %% === Bad:
+        ?_assertNot(verify(MsgBad, MultiSigOrd   , MultiPubKey)),
+        ?_assertNot(verify(MsgBad, MultiSigOrdBad, MultiPubKey)),
+        ?_assertNot(verify(MsgBad, MultiSigOrdBad, MultiPassBad)),
+        ?_assertNot(verify(Msg   , MultiSigOrdBad, MultiPubKey)),
+        ?_assertNot(verify(Msg   , MultiSigOrdBad, MultiPassBad)),
+        ?_assertNot(verify(Msg   , MultiSigOrd   , MultiPassBad)),
+
+        %% === Badder:
+        ?_assertNot(verify(Msg, MultiSigOrd, {multisig, M + 1, N, KeysDigest})),
+        ?_assertNot(verify(Msg, MultiSigOrd, {multisig, M, N + 1, KeysDigest})),
+        ?_assertNot(verify(Msg, MultiSigKeysReord, MultiPubKey)),
+        ?_assertNot(verify(Msg, MultiSigOrdWithOutOfRangeIndex, MultiPubKey))
+    ]
+    ++ MaybeClonedSigTest.
+
+multisig_test_() ->
+    %% TODO Expose list of hash types from multihash lib?
+    HashTypes = [
+        identity,
```

Comment:
I would prefer we not support the identity hash (as it's just the input) and SHA1 is pretty thoroughly broken at this point so we should also probably not allow that.

Maybe we can have a macro listing the unsupported hash types?

Reply:
So use these for negative tests instead?

Reply:
Yeah.